File: generated_code/GeminiPro/Python/085_SplayTree.py
import sys

class SplayTree:
    """
    An implementation of a Splay Tree.

    This class provides a dictionary-like set for integers, supporting
    insert, delete, and search operations. The key feature is the splaying
    operation: whenever a key is accessed (searched, inserted, or deleted),
    the node containing that key (or its parent/successor if the key is not found)
    is moved to the root of the tree. This self-optimizing behavior provides
    amortized O(log n) time complexity for all operations.
    """

    class _Node:
        """A node in the splay tree."""
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation around node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation around node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        if y:
            y.right = x
        x.parent = y

    def _splay(self, node):
        """
        Performs the splaying operation on a node, moving it to the root.
        """
        while node.parent:
            parent = node.parent
            grandparent = parent.parent
            if not grandparent:  # Zig step
                if node == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif node == parent.left and parent == grandparent.left:  # Zig-Zig (left-left)
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif node == parent.right and parent == grandparent.right:  # Zig-Zig (right-right)
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif node == parent.right and parent == grandparent.left:  # Zig-Zag (left-right)
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:  # Zig-Zag (right-left)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def search(self, key):
        """
        Searches for a key in the tree.

        If the key is found, the corresponding node is splayed to the root
        and the method returns True. If the key is not found, the last
        accessed node (the would-be parent) is splayed to the root and
        the method returns False.

        Args:
            key (int): The key to search for.

        Returns:
            bool: True if the key is found, False otherwise.
        """
        node = self.root
        last_node = None
        while node:
            last_node = node
            if key == node.key:
                self._splay(node)
                return True
            elif key < node.key:
                node = node.left
            else:
                node = node.right

        if last_node:
            self._splay(last_node)
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key is not already in the tree, it is inserted and the new
        node is splayed to the root. If the key already exists, the existing
        node is splayed to the root.

        Args:
            key (int): The key to insert.
        """
        parent_node = None
        current_node = self.root
        while current_node:
            parent_node = current_node
            if key < current_node.key:
                current_node = current_node.left
            elif key > current_node.key:
                current_node = current_node.right
            else:
                # Key already exists, splay it and return
                self._splay(current_node)
                return

        new_node = self._Node(key)
        new_node.parent = parent_node

        if not parent_node:
            self.root = new_node
        elif key < parent_node.key:
            parent_node.left = new_node
        else:
            parent_node.right = new_node

        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.

        The node containing the key is first splayed to the root and then
        removed. If the key is not found, the last accessed node is splayed
        to the root.

        Args:
            key (int): The key to delete.
        """
        # First, search for the key. This splays the node (if found)
        # or its would-be parent (if not found) to the root.
        if not self.search(key):
            # Key was not in the tree. search() already splayed the parent.
            return

        # At this point, the node to delete is the root, and we know it exists.
        assert self.root.key == key

        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # The right subtree becomes the new tree
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        else:
            # Disconnect the left subtree
            left_subtree.parent = None
            
            # Find the maximum node in the left subtree
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right

            # Splay this maximum node to be the root of the left subtree.
            # We can do this by temporarily making the left subtree the main tree.
            original_root = self.root
            self.root = left_subtree
            self._splay(max_node)
            
            # Now, self.root is the splayed max_node.
            # Re-attach the original right subtree.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root

File: generated_code/GeminiPro/Java/085_SplayTree.java
/**
 * A complete, self-contained implementation of a Splay Tree for an integer set.
 * Splay trees are self-balancing binary search trees with the additional property
 * that recently accessed elements are quick to access again.
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     * It is a static inner class as it does not need to access instance members of SplayTree.
     */
    private static class Node {
        int key;
        Node left;
        Node right;

        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a right rotation on the subtree rooted at y.
     * @param y The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node rightRotate(Node y) {
        Node x = y.left;
        y.left = x.right;
        x.right = y;
        return x;
    }

    /**
     * Performs a left rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the rotated subtree.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The core splay operation. It brings the node with the given key
     * (or the last accessed node if the key is not found) to the root of the tree.
     * @param currentRoot The root of the current subtree.
     * @param key The key to splay.
     * @return The new root of the splayed tree.
     */
    private Node splay(Node currentRoot, int key) {
        // Base case: root is null or key is at root
        if (currentRoot == null || currentRoot.key == key) {
            return currentRoot;
        }

        // Key is in the left subtree
        if (key < currentRoot.key) {
            // Key is not in the tree, we are done
            if (currentRoot.left == null) {
                return currentRoot;
            }

            // Zig-Zig (Left Left)
            if (key < currentRoot.left.key) {
                currentRoot.left.left = splay(currentRoot.left.left, key);
                currentRoot = rightRotate(currentRoot);
            }
            // Zig-Zag (Left Right)
            else if (key > currentRoot.left.key) {
                currentRoot.left.right = splay(currentRoot.left.right, key);
                if (currentRoot.left.right != null) {
                    currentRoot.left = leftRotate(currentRoot.left);
                }
            }
            // Do second rotation for root
            return (currentRoot.left == null) ? currentRoot : rightRotate(currentRoot);
        }
        // Key is in the right subtree
        else {
            // Key is not in the tree, we are done
            if (currentRoot.right == null) {
                return currentRoot;
            }

            // Zig-Zag (Right Left)
            if (key < currentRoot.right.key) {
                currentRoot.right.left = splay(currentRoot.right.left, key);
                if (currentRoot.right.left != null) {
                    currentRoot.right = rightRotate(currentRoot.right);
                }
            }
            // Zig-Zig (Right Right)
            else if (key > currentRoot.right.key) {
                currentRoot.right.right = splay(currentRoot.right.right, key);
                currentRoot = leftRotate(currentRoot);
            }
            // Do second rotation for root
            return (currentRoot.right == null) ? currentRoot : leftRotate(currentRoot);
        }
    }

    /**
     * Searches for a key in the splay tree.
     * If the key is found, the corresponding node is splayed to become the new root.
     * If not found, the last accessed node on the search path is splayed to the root.
     * @param key The key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        root = splay(root, key);
        if (root.key == key) {
            return key;
        }
        return null;
    }

    /**
     * Inserts a key into the splay tree.
     * If the key already exists, the tree is splayed on that key.
     * If the key is new, it is inserted and becomes the new root of the tree.
     * @param key The key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree for the key. This brings the closest node to the root.
        root = splay(root, key);

        // If the key is already present, do nothing.
        if (root.key == key) {
            return;
        }

        // Create the new node
        Node newNode = new Node(key);

        // If the new key is smaller than the root's key
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        }
        // If the new key is greater than the root's key
        else {
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        // The new node becomes the new root
        root = newNode;
    }

    /**
     * Deletes a key from the splay tree.
     * If the key exists, it is deleted, and the tree is restructured.
     * The parent of the deleted node (or a neighbor) becomes the new root.
     * @param key The key to delete.
     */
    public void delete(int key) {
        if (root == null) {
            return;
        }

        // Splay the tree for the key.
        // If the key is in the tree, it will become the root.
        root = splay(root, key);

        // If the key is not at the root, it's not in the tree.
        if (root.key != key) {
            return;
        }

        // Now, the node to delete is the root.
        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            // If there's no left child, the right child becomes the new root.
            root = rightSubtree;
        } else {
            // Splay the left subtree to bring its maximum element to its root.
            // This maximum element is the predecessor of the key we are deleting.
            leftSubtree = splay(leftSubtree, key);

            // The original right subtree becomes the right child of this new root.
            leftSubtree.right = rightSubtree;
            root = leftSubtree;
        }
    }
}